### PR TITLE
Leaderboard integation with experiments

### LINKED ISSUE #1211: Figure out an approach for adding experiments to the leaderboard
We currently do not have a good approach to adding experiment runs to the MTEB leaderboard e.g. experiment with the influence of hyperparameters. Such as using no-prompt for prompt-based models or changing the embedding size.

A solution to implement is to run it with a unique model_name, but there is currently no documentation on how one would need to do that.

---

edit: a potentially better solution is to add experiments. Where we add a layer to the visualization:

```
model_name
| - revision1
|    | - task1.json
     | - ...
     | - experiments
     |    | - exp_name1
     |    |    | -  task1.json
     |    |    | - ...
     |    | - exp_name1
     |    |    | -  task1.json
     |    |    | - ...
```

or potentially a slightly more consistent structure:

```
model_name
| - revision1
     | - default
     |    | -  task1.json
     |    | - ...
     | - exp_name1
     |    | -  task1.json
     |    | - ...
```

Experiment names could be e.g. no-instruct/instruct (whether you use instructions) or emb_size=256

To run a model as an experiment we could do:

```
model = MyModel(**custom_kwargs)
results = mteb.MTEB(model, experiment_name = "no-instruct")
```

### PR DESCRIPTION
**Pull request by Samoed**
Close https://github.com/embeddings-benchmark/mteb/issues/1211
Frontend https://github.com/embeddings-benchmark/leaderboard-frontend/pull/4
Reapplying changes from https://github.com/embeddings-benchmark/mteb/pull/4779

### CODE REVIEW

File: mteb/results/benchmark_results.py [hunk 1]
```diff
@@ -458,10 +473,19 @@ def _build_pre_agg_df(self, include_model_revision: bool) -> pd.DataFrame | None
...
         for model_result in bench_results:
             mn = model_result.model_name
             mr = model_result.model_revision
+            mm = model_result.model_meta
+            exp_kwargs = (
+                dict(mm.experiment_kwargs)
+                if mm is not None and mm.experiment_kwargs
+                else None
+            )
+            meta_dict = mm.to_dict() if mm is not None and exp_kwargs else None
```
Comment (Samoed) [inline mteb/results/benchmark_results.py:488]: I added model meta to the results to load it for experiments for setups like jina v4 when based on experiment paramters model meta changes a bit (e.g. from dense to late interaction). But with this approach model meta that stored in results can be old and not match with what we have in code, so I'm not sure if it good approach 

Image

File: mteb/api/adapters.py [hunk 2]
```diff
@@ -84,6 +88,39 @@ def model_meta_to_schema(
     return cached.model_copy(update={"zero_shot_pct": int(zero_shot_pct)})
 
 
+def run_model_meta_to_schema(
+    run_meta: dict[str, Any],
+    *,
+    zero_shot_pct: int | None = None,
+) -> ModelMetaSchema | None:
+    """`ModelMetaSchema` built straight from one *run's own* `model_meta.json`.
```
Comment (KennethEnevoldsen) [inline mteb/api/adapters.py:96]: Ahh but we I am not at all sure we trust the model_meta.json except for the experiment kwargs (the rest can change)

Comment (KennethEnevoldsen) [reply mteb/api/adapters.py:96]: I suspect that we need to handle a merge here

Comment (KennethEnevoldsen) [reply mteb/api/adapters.py:96]: (e.g. training data)

File: mteb/api/aggregators.py [hunk 3]
```diff
@@ -244,6 +244,49 @@ async def build_benchmark_summary(  # noqa: PLR0914
...
+    # ``experiments`` column so variant rows can surface the exact ablation
+    # dict that produced them. Empty when the parquet pre-dates the
+    # experiments work.
+    #
+    # Also carries each variant's own `model_meta.json` dict (the
+    # ``model_meta`` column) so its row can build a `ModelMetaSchema` from
+    # that run's *actual* metadata instead of the static MODEL_REGISTRY
+    # entry — an ablation can change more than the kwarg it's named after
+    # (e.g. jinaai/jina-embeddings-v4's `vector_type=multi_vector`
+    # experiment runs `late-interaction`, not the base model's `dense`).
```
Comment (KennethEnevoldsen) [inline mteb/api/aggregators.py:258]: this is very long and I am not quite sure what it is trying to say -- I suspect it is ai generated an probably can be rewritten to be more precise

File: mteb/benchmarks/_create_table.py [hunk 4]
```diff
@@ -252,24 +252,63 @@ def _get_embedding_size(embed_dim: int | Sequence[int] | None) -> int | None:
     return None
 
 
+_VARIANT_ID_COL = "_variant_id"
+
+
+def _ensure_variant_id(pl_df: pl.DataFrame) -> pl.DataFrame:
+    """Add a stable ``_variant_id`` Utf8 column derived from ``experiments``.
```
Comment (KennethEnevoldsen) [inline mteb/benchmarks/_create_table.py:259]: this column seems a bit unclear to me? Is it just the model id (including experiments)

variant_id seems very vague to me

Comment (KennethEnevoldsen) [reply mteb/benchmarks/_create_table.py:259]: or is it just the id for the experiment? in which case why not use `experiment_id`?

File: mteb/results/benchmark_results.py (hunk 1 above)
Comment (KennethEnevoldsen) [reply mteb/results/benchmark_results.py:488]: Hmm maybe we need to merge. Essentially state what we trust from the code and what we allow to change.

e.g. license does not change, training data should probably just include the joint set (I could imagine experiment fitting on data).

We probably need some tests in the results repo to validate the model_meta.json if we plan to use them like this (e.g. checking that license match)

Comment (KennethEnevoldsen) [reply mteb/results/benchmark_results.py:488]: Or we should be clear what an experiment can an can't change (and then merge on the rest)

File: mteb/results/benchmark_results.py [hunk 5]
```diff
@@ -256,16 +256,21 @@ def join_revisions(self) -> BenchmarkResults:
         Returns:
             A new BenchmarkResults object with the revisions joined.
         """
+        _experiment_sentinel = ""
```
Comment (KennethEnevoldsen) [inline mteb/results/benchmark_results.py:259]: I am not sure why this is needed (also sting seems like an incorrect replacement for experiment name, I would argue that at non-experiment should have None)

File: mteb/results/benchmark_results.py
Comment (KennethEnevoldsen) [inline mteb/results/benchmark_results.py:1]: would love a test on these changes

File: mteb/cache/result_cache.py
Comment (KennethEnevoldsen) [inline mteb/cache/result_cache.py:1]: would love a test on these changes

**Review by KennethEnevoldsen — COMMENTED** [review]
initial review, I think I am mostly missing info - can you add a PR comment describing the PR?

Looks like we also need to update API docs?

File: mteb/results/benchmark_results.py (hunk 1 above)
Comment (Samoed) [reply mteb/results/benchmark_results.py:488]: Yes, but overall this complicates pipeline and I'm not sure if we really want to add, because this is really niche use case. I think only embed dim and dtype can be changes, but experiment "tag" would give more visibility. I think we have only one model with such case like jina which probably just rename arg instead and not to save `model_meta` in results

File: mteb/api/aggregators.py (hunk 3 above)
Comment (Samoed) [reply mteb/api/aggregators.py:258]: Yes, I'll update a bit later

File: mteb/results/benchmark_results.py (hunk 1 above)
Comment (KennethEnevoldsen) [reply mteb/results/benchmark_results.py:488]: So what are you suggesting?

Comment (Samoed) [reply mteb/results/benchmark_results.py:488]: I'm not sure if we want to cover cases like this. Ovearall, I think we can scope that `model_type` and `embed_dim` can change in experiments to be loaded, but I think would be just easier to ignore this, because embed dim would be more visilble as "eximent" tag and `model_type` change is very rare case which we might want to ignore it (overall I know only jina-v4 and in theory bge-m3 with this case) and update experiment kwarg a bit to be more clear on ledaerboard. What do you think?

Comment (KennethEnevoldsen) [reply mteb/results/benchmark_results.py:488]: Hmm I could just imagine it to be a more common case in the future (one model to do more things) -- but agree I would probably consider reducing experiments to only include embed-dim and dtype for now (then we can expand it in the future). Does that work?

Comment (Samoed) [reply mteb/results/benchmark_results.py:488]: Yes, I think that would work

**Comment by KennethEnevoldsen** [comment]
@Samoed I assume this is still ongoing - are you waiting for anything on my end?

**Comment by Samoed** [comment]
Not yet, I'll update this soon

**Comment by KennethEnevoldsen** [comment]
@Samoed will remove stale here, I assume we don't want this closed

**Comment by Samoed** [comment]
Yeah, I'll try to get back to my pr tomorrow. Hope, there won't be much a lot of pr to review

**Comment by KennethEnevoldsen** [comment]
I am trying to get back to some reviewing now that I see that the ICLR paper will make it (the gsm one). I was a bit worried it wouldn't so had to work a bit overtime on that

**Comment by Samoed** [comment]
I haven't look into gym at all. Maybe I need to look at it a bit too

**Comment by KennethEnevoldsen** [comment]
No the math one (I haven't looked at gym either)

File: mteb/cache/result_cache.py [hunk 6]
```diff
@@ -718,13 +722,16 @@ def _load_from_cache(
                 {cache_path}/leaderboard/{cache_filename}.
             rebuild: If True, force a full rebuild from the results repository, bypassing any
                      pre-computed JSON cache.
+            save_rebuilt_to_disk: Whether a full rebuild (strategy 3, or
+                forced via ``rebuild=True``) persists its result to
+                ``cache_path``. See ``_rebuild_from_full_repository``.
```
Comment (KennethEnevoldsen) [inline mteb/cache/result_cache.py]: I am not sure what this means

File: mteb/cache/result_cache.py [hunk 7]
```diff
@@ -941,43 +964,49 @@ def get_task_names(
...
-                f"model_meta.json not found in {revision_path}, extracting model_name and revision from the path"
-            )
+        model_meta_path = revision_path / "model_meta.json"
+        if not model_meta_path.exists():
             if _EXPERIMENTS_FOLDER_NAME in revision_path.parts:
-                logger.debug(
-                    f"Path {revision_path} contains an experiment folder, extracting model_name and revision accordingly"
+                # Warning (not debug): drops every task file under this
+                # folder, and only logs once per folder since this method
+                # is lru_cache'd.
```
Comment (KennethEnevoldsen) [inline mteb/cache/result_cache.py]: I get the last part but not the first part here

File: mteb/cache/result_cache.py [hunk 8]
```diff
@@ -1402,14 +1431,18 @@ def load_results(
...
         for path in paths:
             task_result = TaskResult.from_disk(path)
 
             if only_main_score:
                 task_result = task_result.only_main_score()
-            model_name, revision, experiment_name = (
-                self._get_model_name_and_revision_from_path(path.parent)
-            )
+            # `_get_model_name_and_revision_from_path` already warns (once
+            # per folder, lru_cache'd) when it returns None.
```
Comment (KennethEnevoldsen) [inline mteb/cache/result_cache.py]:
```suggestion
```

File: mteb/models/model_meta.py [hunk 9]
```diff
@@ -1850,6 +1853,19 @@ def _collect_similar_tasks(dataset: str, visited: set[str]) -> set[str]:
     return similar
 
 
+def _has_meaningful_value(value: Any) -> bool:  # noqa: ANN401 -- checks arbitrary kwarg values
```
Comment (KennethEnevoldsen) [inline mteb/models/model_meta.py]:
```suggestion
def _has_meaningful_value(value: Any) -> bool:  # noqa: ANN401
```

**Review by KennethEnevoldsen — COMMENTED** [review]
mostly minor stuff

File: mteb/cache/result_cache.py (hunk 6 above)
Comment (Samoed) [reply mteb/cache/result_cache.py]: `save_rebuilt_to_disk=False` won't save `__cached_results.json` in cache. I had some validation problems with it, and I don't think it was possible to load it before pr. Overall, I don't think it was really used at any time

File: mteb/cache/result_cache.py (hunk 7 above)
Comment (Samoed) [reply mteb/cache/result_cache.py]: No results in the `experiments` folder will be loaded if the `model_meta` folder does not exist
